knative/logs.go

```diff
@@ -0,0 +1,119 @@
+package knative
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sync"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"knative.dev/kn-plugin-func/k8s"
+)
+
+// GetKServiceLogs will get logs of Knative service.
+//
+// It will do so by gathering logs of user-container of all affiliated pods.
+// In addition, filtering on image can be done so only logs for given image are logged.
+// The image must be the digest format since pods of Knative service use it.
+//
+// This function runs as long as the passed context is active (i.e. it is required cancel the context to stop log gathering).
+func GetKServiceLogs(ctx context.Context, namespace, kServiceName, image string, since *time.Time, out io.Writer) error {
```

Comment:
@rhuss @dsimansk is there any off the shelf solution that works?
I tried `stern` and `kn-plugin-service-log` neither of them is good enough.
They are not displaying error `exec /bin/sh: exec format error` when deploying image with bad arch.
I don't know why.

Reply:
https://github.com/knative-sandbox/kn-plugin-func/blob/e041bc84ebe4accc7e099cc45107269164e94066/k8s/logs.go#L13

Reply:
@zroubalik I've seen that one but I wanted to gather logs in streaming/continuous way not ex post. Also I am gathering logs from multiple pods.

Reply:
>I wanted to gather logs in streaming/continuous way not ex post.

btw I think that would be also useful for pipeline build with `-v` flag.

Reply:
>I wanted to gather logs in streaming/continuous way not ex post.

I want to print error ASAP, not after 2 minutest of futile CrashLoopBackOff-Error loop.

Reply:
> > I wanted to gather logs in streaming/continuous way not ex post.
> 
> btw I think that would be also useful for pipeline build with `-v` flag.

That would be nice.

I am not sure if this is the canonical way of reading logs, but `tkn` (Tekton cli) has this capability and it's quite nice, might be worth checking that out: 
https://github.com/tektoncd/cli/blob/4f6674193b2b6f41c32e14b1e0978d7e1b66c39a/pkg/pods/container.go#L88